main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ModuleCompleter.scala

```diff
@@ -17,14 +17,46 @@ package ca.uwaterloo.flix.api.lsp.provider.completion
 
 import ca.uwaterloo.flix.api.Flix
 import ca.uwaterloo.flix.api.lsp.Index
-import ca.uwaterloo.flix.language.ast.TypedAst
+import ca.uwaterloo.flix.language.ast.{Symbol, TypedAst}
 import ca.uwaterloo.flix.api.lsp.provider.completion.Completion.ModCompletion
+import ca.uwaterloo.flix.api.lsp.provider.completion.CompletionUtils.generateModSymAndSubwordFromContext
+import ca.uwaterloo.flix.language.ast.Symbol.ModuleSym
 
 object ModuleCompleter extends Completer {
   /**
     * Returns a List of ModCompletion for modules.
     */
   override def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root, delta: DeltaContext): Iterable[ModCompletion] = {
-    root.modules.keys.map(mod => ModCompletion(mod))
+    val (modSym, subWord) = generateModSymAndSubwordFromContext(context)
+
+    getModuleCompletion(modSym, subWord)
+  }
+
+  /**
+    * Get all module completions
+    */
+  private def getModuleCompletion(modSym: Symbol.ModuleSym, subWord: String)(implicit root: TypedAst.Root): Iterable[ModCompletion] = {
+    // Use fqn to lookup in modules and get all syms
+    val symsInModule = root.modules.getOrElse(modSym, Nil)
+
+    // Get all modules that matches word
+    val validModsInModule = symsInModule.collect {
+      case sym: ModuleSym if matchesMod(sym, subWord) => sym
+    }
+
+    // Generate completions
+    validModsInModule.map(mod => ModCompletion(mod))
+  }
+
+  /**
+    * Checks if the last elem of the nameSpace matches the subWord (the word the user is currently typing).
+    *
+    * @param mod     the moduleSym.
+    * @param subWord the subWord provided by the user.
+    * @return        true, if the subWord matches the last part of nameSpace, false otherwise.
+    */
+  private def matchesMod(mod: ModuleSym, subWord: String): Boolean = {
+    val lastNs = mod.ns.takeRight(1)(0)
```

Comment:
Can this not go index out of bounds? You need to check its not the empty list.

Reply:
I've added a guard